### PR TITLE
feat(superuser): Disable default su access in login flow

### PR DESCRIPTION
**Pull request by maxiuyuan**
Users will not longer be able to gain superuser access through the login flow and only through the superuser modal (future pr with enable it for _admin, wont merge until that is finished).

This pr removes the `enable_superuser` out of the receiver and thus su access is only given through the `PUT` method on `auth_index` endpoint.
Also changed `auth_config` and `auth_login` to not check if the user is an active superuser since we no longer need to redirect the user to the login flow to gain su access. Also removed the tests that accompanied the removed code

[notion doc for more detail of each pr with action diagram and user story](https://www.notion.so/sentry/Better-Accountability-for-Superuser-access-PRs-0854562e6b4a4a1aa7744a7f43b82a54)

### CODE REVIEW

File: src/sentry/auth/superuser.py [hunk 1]
```diff
@@ -135,11 +135,13 @@ def get_session_data(self, current_datetime=None):
                     extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
                 )
             return
+        # don't want to spam the logs since superuser session is not activated on log in
```
Comment (RyanSkonnord) [inline src/sentry/auth/superuser.py:138]: Nit: this comment should be flush with `if getattr(self, "token", None)`.

File: src/sentry/auth/superuser.py [hunk 2]
```diff
@@ -304,17 +305,20 @@ def set_logged_in(self, user, current_datetime=None):
...
-        logger.info(
-            "superuser.logged-in",
-            extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": user.id},
-        )
+            logger.info(
+                "superuser.logged-in",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": user.id},
+            )
+
+        except AttributeError:
```
Comment (RyanSkonnord) [inline src/sentry/auth/superuser.py:320]: What is this anticipating? `AttributeError` is quite broad.

If this is meant to proceed even if one of the `logger.info` calls fails, that's good, but as written it would cause us to skip calling `_set_logged_in` if the first `logger.info` chokes.

File: src/sentry/testutils/cases.py [hunk 3]
```diff
@@ -203,6 +203,12 @@ def make_request(self, user=None, auth=None, method=None, is_superuser=False):
         request.superuser = Superuser(request)
         if is_superuser:
             # XXX: this is gross, but its a one off and apis change only once in a great while
+            request._body = json.dumps(
+                {
+                    "superuserAccessCategory": "Edit organization settings",
+                    "superuserReason": "Edit organization settings",
+                }
+            )
             request.superuser.set_logged_in(user)
```
Comment (RyanSkonnord) [inline src/sentry/testutils/cases.py:212]: This change makes the scope of the comment unclear; before it was referring to `set_logged_in`, but now it's unclear if it also refers to hacking on `request._body`. (To which the comment also would kind of apply, actually.)

```suggestion
            request._body = json.dumps(
                {
                    "superuserAccessCategory": "Edit organization settings",
                    "superuserReason": "Edit organization settings",
                }
            )

            # XXX: this is gross, but its a one off and apis change only once in a great while
            request.superuser.set_logged_in(user)
```

File: src/sentry/testutils/cases.py [hunk 4]
```diff
@@ -243,6 +249,12 @@ def login_as(
             # XXX(dcramer): we're calling the internal method to avoid logging
             request.superuser._set_logged_out()
         elif request.user.is_superuser and superuser:
+            request._body = json.dumps(
```
Comment (RyanSkonnord) [inline src/sentry/testutils/cases.py:252]: Messing with an underscore-prefixed attribute should be a last resort, even in a testing context. Could we look into alternatives?

My first pass would be to find the code that expects these keys and extract a function that can be mocked out in the affected tests.

File: tests/sentry/auth/test_superuser.py [hunk 5]
```diff
@@ -71,10 +72,19 @@ def build_request(
     def test_ips(self):
         user = User(is_superuser=True)
         request = self.make_request(user=user)
+        request._body = json.dumps(
+            {
+                "superuserAccessCategory": "Edit organization settings",
+                "superuserReason": "Edit organization settings",
+            }
+        )
```
Comment (RyanSkonnord) [inline tests/sentry/auth/test_superuser.py:80]: Doesn't `make_request` now do this with `is_superuser=True`? Is there some reason I'm missing that we don't want to call `make_request` that way?

File: tests/sentry/auth/test_superuser.py [hunk 6]
```diff
@@ -71,10 +72,19 @@ def build_request(
...
         request.META["REMOTE_ADDR"] = "10.0.0.1"
 
         # no ips = any host
         superuser = Superuser(request, allowed_ips=())
+        request.session[SsoSession.django_session_key(superuser.org_id)] = {
+            "ts": timezone.now().timestamp()
```
Comment (RyanSkonnord) [inline tests/sentry/auth/test_superuser.py:86]: Is this just a placeholder?

If there's any chance of the timestamp affecting test behavior, I'd rather replace it with a constant or something derived from `self.current_datetime`.

File: tests/sentry/auth/test_superuser.py (hunk 5 above)
Comment (maxiuyuan) [reply tests/sentry/auth/test_superuser.py:80]: It only does that if `is_superuser=True` is passed into `make_request` but unfortunately there are unwanted behaviours that `make_request` does if `is_superuser=True` there are examples of other tests that test superuser functionalities that call `make_request` without `is_superuser=True` to avoid the unwanted behaviours

File: src/sentry/auth/superuser.py (hunk 2 above)
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:319]: This is anticipating if there are any missing information in request.user. If that happens we want to skip `_set_logged_in` and the subsequent  `superuser.logged-in` log

File: src/sentry/testutils/cases.py (hunk 4 above)
Comment (JoshFerge) [inline src/sentry/testutils/cases.py:252]: I would consider adding a setting to toggle checking SU ACCESS reason, as it would be useful not to have to provide this mocked data, and additionally in development it is probably worth hiding for better dx anyways.

File: src/sentry/auth/superuser.py (hunk 2 above)
Comment (JoshFerge) [reply src/sentry/auth/superuser.py:340]: @maxiuyuan how would that ever happen?

Comment (JoshFerge) [reply src/sentry/auth/superuser.py:340]: is there a test case that covers this?

File: src/sentry/auth/superuser.py [hunk 7]
```diff
@@ -139,10 +139,11 @@ def get_session_data(self, current_datetime=None):
                 )
             return
         elif not data:
-            logger.warning(
-                "superuser.missing-session-data",
-                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
-            )
+            if getattr(self, "token", None):
```
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:142]: why was this line added?

File: tests/sentry/web/frontend/test_auth_login.py [hunk 8]
```diff
@@ -252,12 +252,22 @@ def test_redirects_already_authed_non_superuser(self):
             resp = self.client.get(self.path)
             self.assertRedirects(resp, "/organizations/new/")
 
-    def test_doesnt_redirect_already_authed_superuser(self):
```
Comment (maxiuyuan) [inline tests/sentry/web/frontend/test_auth_login.py:255]: The behaviour of the view has changed as superuser access isn't granted upon login, so we no longer need to redirect a non active superuser

File: tests/sentry/api/endpoints/test_auth_config.py [hunk 9]
```diff
@@ -39,7 +39,7 @@ def test_superuser_is_not_redirected(self):
         response = self.client.get(self.path)
 
         assert response.status_code == 200
-        assert "nextUri" not in response.data
```
Comment (maxiuyuan) [inline tests/sentry/api/endpoints/test_auth_config.py:42]: As we are no longer granting superuser access upon login, we don't need to redirect the user back to sso login and they can continue with the normal flow

**Review by RyanSkonnord — APPROVED** [review]
